File: func2.js
//  func2

const Box = x => ({
    map: f => Box(f(x)),
    fold: f => f(x),
    inspect: () => `Box(${x})`
})

// const moneyToFloat = str =>
//     parseFloat(str.replace(/\$/g, ''))

const moneyToFloat = str =>
    Box(str)
    .map(s => s.replace(/\$/g, ''))
    .map(t => parseFloat(t))

// const percentToFloat = str => {
//     const replaced = str.replace(/\%/g, '')
//     const number = parseFloat(replaced)
//     return number * 0.01
// }

const percentToFloat = str =>
    Box(str)
    .map(s => s.replace(/\%/g, ''))
    .map(t => parseFloat(t))
    .map(number => number * 0.01)


// const applyDiscount = (price, discount) => {
//     const cost = moneyToFloat(price)
//     const savings = percentToFloat(discount)
//     return cost - cost * savings
// }

const applyDiscount = (price, discount) =>
    moneyToFloat(price)
    .fold(cost => 
        percentToFloat(discount)
        .fold(savings =>
            cost - cost * savings))


const result = applyDiscount('$5.00', '20%')

console.log(result)

File: func2.py
# func2

class Box:
    def __init__(self, x):
        self.x = x
        self.map = lambda f: Box(f(x))
        self.fold = lambda f: f(x)
    
    def __repr__(self):
        return 'Box({})'.format(self.x)


def money_to_float(cost):
    return Box(cost) \
        .map(lambda s: s.replace("$", "")) \
        .map(lambda t: float(t))

def percent_to_float(discount):
    return Box(discount) \
        .map(lambda s: s.replace("%", "")) \
        .map(lambda t: float(t)) \
        .map(lambda number: number * 0.01)

def applyDiscount(price, discount):
    return money_to_float(price) \
        .fold(lambda cost: percent_to_float(discount).fold(lambda savings: cost - cost * savings))

result = applyDiscount('$5.00', '20%')

# print("cost is: {}".format(money_to_float('$5.00')))
# print("discount is: {}".format(percent_to_float('20%')))

print(result)
